### PR TITLE
Typescript first

### PR DESCRIPTION
**Pull request by bidoubiwa**
# Typescript conversion

## Types update
Types have been improved and are compatible with instant search types.

## Linting update

All envs and playgrounds are linted based on the root eslint config file.

## Jest updates

as in meilisearch-js meilisearch/meilisearch-js#760 the jest jsdom had to be updated for the tests to succeed in node 10. https://github.com/meilisearch/instant-meilisearch/issues/255

## Testing

Add build before tests for env tests using jest to succeed.
Add build of playgrounds in CI to determine if they correctly extract instant-meilisearch.
Add instantsearch.js as a dev dependency as it is used for testing purposes
Add some nodejs tests that tries instantsearch compatibility with instant-meilisearch
Add a typescript 

## Playgrounds

- Update playgrounds so that hot-reload also works on typescript files
- Remove sub-dir eslint configs as everything is handled by root eslint config file

## Naming

Methods inside instant-meilisearch function have been renamed for more clarity

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,64 +1,173 @@
...
+      parserOptions: {
+        ecmaFeatures: {
+          jsx: true,
         },
-        singleline: {
-          delimiter: 'semi', // 'semi' or 'comma'
-          requireLast: false,
+        ecmaVersion: 2018,
+        sourceType: 'module',
+        projectFolderIgnoreList: ['dist', 'playgrounds/vue'],
```
Comment (curquiza) [inline .eslintrc.js:25]:
```suggestion
```

File: .eslintrc.js [hunk 2]
```diff
@@ -1,64 +1,173 @@
...
+    },
+    {
+      /*
+       * JS
+       */
+      files: [
+        '**/*.js',
+        'scripts/**/*.js',
+        'playgrounds/**/*.js',
+        'tests/env/**/*.js',
```
Comment (curquiza) [inline .eslintrc.js:44]:
```suggestion
```

File: .eslintrc.js [hunk 3]
```diff
@@ -1,64 +1,173 @@
...
+      },
+      extends: [
+        'standard',
+        'plugin:prettier/recommended',
+        'eslint:recommended',
+      ],
+      plugins: ['jest'],
+      parserOptions: {
+        ecmaVersion: 11,
+        // sourceType: 'module',
```
Comment (curquiza) [inline .eslintrc.js:68]:
```suggestion
```

File: .eslintrc.js [hunk 4]
```diff
@@ -1,64 +1,173 @@
...
+        commonjs: true,
+        es6: true,
+      },
+      extends: [
+        'plugin:prettier/recommended',
+        'prettier/vue',
+        'plugin:vue/essential',
+        'plugin:vue/base',
+      ],
+      // extends: ['plugin:vue/essential', 'prettier/vue'],
```
Comment (curquiza) [inline .eslintrc.js:89]:
```suggestion
```

File: src/index.ts [hunk 5]
```diff
@@ -93,112 +16,160 @@ export default function instantMeiliSearch(
...
+      )
+      const ISHits = this.transformToISHits(hits, instantSearchParams)
+      const ISResponse = {
         index: indexUid,
-        hitsPerPage: this.hitsPerPage!,
-        facets,
-        exhaustiveFacetsCount,
+        hitsPerPage: this.hitsPerPage,
+        ...(facets && { facets }),
+        disjunctiveFacets: [],
```
Comment (curquiza) [inline src/index.ts:129]:
```suggestion
```

File: src/types.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import * as MStypes from 'meilisearch'
+import * as IStypes from './instantsearchTypes'
+export * as InstantsearchTypes from './instantsearchTypes'
```
Comment (curquiza) [inline src/types.ts:3]:
```suggestion
export * as InstantSearchTypes from './instantsearchTypes'
```

**Review by curquiza — APPROVED** [review]
Huge work! thanks @bidoubiwa!

I let your merge when you are ready!

**Comment by bidoubiwa** [comment]
bors merge

**Pull request merged**